src/Tools/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Codefix.Testing/BaseCodeFixTest.cs

```diff
@@ -0,0 +1,589 @@
+                if (analyzerDiagnostics.Length == 0)
+                {
+                    break;
+                }
+
+                if (!AreDiagnosticsDifferent(analyzerDiagnostics, previousDiagnostics))
+                {
+                    break;
+                }
+
+                if (--numberOfIterations < -1)
+                {
+                    throw new Exception("The upper limit for the number of code fix iterations was exceeded");
+                }
+
+                previousDiagnostics = analyzerDiagnostics;
+
+                done = true;
+                var anyActions = false;
+                foreach (var diagnostic in analyzerDiagnostics)
+                {
+                    if (!codeFixProvider.FixableDiagnosticIds.Contains(diagnostic.Id))
+                    {
+                        // do not pass unsupported diagnostics to a code fix provider
+                        continue;
+                    }
+
+                    var actions = new List<CodeAction>();
+                    var context = new CodeFixContext(project.GetDocument(diagnostic.Location.SourceTree), diagnostic, (a, d) => actions.Add(a), cancellationToken);
+                    await codeFixProvider.RegisterCodeFixesAsync(context).ConfigureAwait(false);
+
+                    if (actions.Count > 0)
+                    {
+                        anyActions = true;
+
+                        var fixedProject = await ApplyFixAsync(project, actions.ElementAt(codeFixIndex.GetValueOrDefault(0)), cancellationToken).ConfigureAwait(false);
+                        if (fixedProject != project)
+                        {
+                            done = false;
+
+                            project = await RecreateProjectDocumentsAsync(fixedProject, cancellationToken).ConfigureAwait(false);
+                            break;
+                        }
+                    }
+                }
+
+                if (!anyActions)
+                {
+                    Verify.True(done, "Expected to be done executing actions.");
+
+                    // Avoid counting iterations that do not provide any code actions
+                    numberOfIterations++;
+                }
+            }
+            while (!done);
+
+            if (expectedNumberOfIterations >= 0)
+            {
+                if (expectedNumberOfIterations != (expectedNumberOfIterations - numberOfIterations))
+                {
+                    throw new Exception($"Expected '{expectedNumberOfIterations}' iterations but found '{expectedNumberOfIterations - numberOfIterations}' iterations.");
+                }
+            }
+
+            return project;
+        }
+
+        /// <summary>
+        /// Implements a workaround for issue #936, force re-parsing to get the same sort of syntax tree as the original document.
```

Comment:
:bulb: Could expand this to a full link to https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/936

Reply:
Please remove this workaround - this basically hides bugs in Roslyn. See https://github.com/dotnet/roslyn-analyzers/pull/1525 and discussion in it. Majority of Roslyn team prefers not to take this approach.

Reply:
:+1:

Reply:
❗️ We may need a central hook for this for StyleCop Analyzers. I will experiment to see if removing it would be a hard adoption blocker.

Reply:
@mavasani I reviewed the use of this method today, and it is essential for correct behavior of the test framework. The code fix verifier runs multiple iterations where diagnostics are obtained and fixes applied, similar to what would occur if a user would experience if they applied multiple code fixes as distinct operations in sequence. The editor applies this normalization step automatically; the test does not inherit this functionality automatically and must manually ensure the state is correct at the beginning of each iteration.

This code is a correctness requirement, not a convenience requirement. If it's removed, it would need to be replaced with other code that retains the correctness aspect of the original.

Reply:
> we should default the tests to an accurate representation of the behavior a user would observe

I disagree, as this hides bugs in Roslyn APIs, and we have already concluded in https://github.com/dotnet/roslyn-analyzers/pull/1525 that this workaround is undesirable by default.
Alternatively, we can include a hook that repos such as StyleCopAnalyzers can turn on to invoke this workaround to do the round tripping. As concluded in https://github.com/dotnet/roslyn-analyzers/pull/1525, the workaround should **not** be applied by _default_ as it defeats the purpose of API unit testing, one could as well write an IDE integration test if the behavior in VS IDE needs to be tested.

Reply:
> I disagree, as this hides bugs in Roslyn APIs, and we have already concluded in dotnet/roslyn-analyzers#1525 that this workaround is undesirable by default.

It's possible the usage here in this PR is a different context from the one you referenced. At minimum, the relation to the internal behavior of Roslyn APIs is substantially different. The behavior as implemented from the perspective of an external analyzer package is instrumental for building confidence in the behavior of the analyzer. I have no objection to continuing the discussion separately from this pull request, but I don't think we should deviate on this facet from the behavior proven to work in the initial iterative development.

Reply:
@mavasani How about a compromise default: include the normalization step, but by default have an assertion that the normalization step does not change the form of the tree (i.e. the normalization transform is an identity transform). This provides the guarantees of valid input to the next iteration while also validating the previous step produced an expected tree.

Reply:
The text of a compilation unit uniquely determines the syntax tree which is passed to analyzers and provided as input to code fixes. The Visual Studio IDE and the command line compiler enforce ensure this invariant is met in user scenarios. Analyzers and code fixes should be executed against this form for the optimum ability to detect incorrect behavior.

The implemention in this pull request assumes that the host is responsible for ensuring tree-based transformations produce inputs to the next diagnostic phase that adhere to the invariants for text-based input, which is consistent with the hosts mentioned above. If the host is not responsible for ensuring trees meet compiler invariants, we can remove the silent normalization. However, the invariants themselves remain. 

The normalization step ensures the analyzers are provided correct trees for analysis. The assertion in question ensures that code fixes produce the normalized form as a direct output. So we end up with one of the following:

1. Normalization is the responsibility of the host, and we silently provide the behavior while acting as a host in incremental analysis scenarios
2. Normalization is the responsibility of the code fix, and we include an assertion that the requirement is met

In practice, I believe the first case is a reasonable expectation of third party analyzer authors, for the following reasons:

1. Normalization is difficult to ensure under fix all conditions
2. Normalization is difficult to ensure for code fixes that manipulate trivia
3. It is only possible to observe a failure of the host environment to normalize the tree if a sequence of strictly tree-based transformations is applied without ever round-trip to a text-based form

The third item is an easily avoided edge case that makes the work of everyone else easier. I can understand an argument that the build-in code fixes should produce normalized form outputs, but it's not a reasonable expectation for third party analyzer authors who have enough difficulty dealing with fixes when the host *is* providing normalization. If in fact normalization is an expectation of the host, code fix tests should not fail by default when the fix both 1) produced the correct text output and 2) met the postcondition for a code fix.

Reply:
The fundamental question is:

:question: Who is responsible for ensuring the syntax tree after applying a code fix is the same as the syntax tree that *would* be produced by the compiler after parsing the text representation of the tree?

Reply:
Thanks for clarifying your views. So basically, your proposal is to perform normalization to validate the syntax trees (and not just the underlying texts) coming out from a code fix are identical to the ones that will be generated after round tripping and re-parsing by the compiler? If so, that sounds a reasonable validation step. We should then change the validation methods to be as follows:

```
        private static async Task VerifyRecreateProjectDocumentsAsync(Project project, CancellationToken cancellationToken)
        {
            foreach (var documentId in project.DocumentIds)
            {
                var document = project.GetDocument(documentId);
                await VerifyRecreateDocumentAsync(document, cancellationToken).ConfigureAwait(false);
            }
             return project;
        }

        private static async Task VerifyRecreateDocumentAsync(Document document, CancellationToken cancellationToken)
        {
            var newText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
            newText = newText.WithChanges(new TextChange(new TextSpan(0, 0), " "));
            newText = newText.WithChanges(new TextChange(new TextSpan(0, 1), string.Empty));
            var newDocument = document.WithText(newText);
            // Add code to assert that document.SyntaxTree and newDocument.SyntaxTree are identical
        }
```

Reply:
I would prefer to file a follow-up issue to add this validation primarily because the message that appears if the `Assert.Equal` in my example fails is useless (it prints the text of the trees, but we already know the text will match).

Reply:
I am fine with the above logic with couple of minor edits, which emphasize that the round tripping is an assertion of a fact that ideally **must** be respected by each code transformation provider:
1. `ValidateCodeFixTreeIsNormalized` is `true` by default.
2. `project` is updated only if `ValidateCodeFixTreeIsNormalized = false`, i.e. the non-default code path.

```diff
private static async Task<Project> RecreateProjectDocumentsAsync(Project project, CancellationToken cancellationToken)
 {
     foreach (var documentId in project.DocumentIds)
     {
         var document = project.GetDocument(documentId);
+        var initialTree = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
         document = await RecreateDocumentAsync(document, cancellationToken).ConfigureAwait(false);
+        var recreatedTree = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+        if (ValidateCodeFixTreeIsNormalized)
+        {
+            Assert.Equal(initialTree, recreatedTree);
+        }
+        else
+        {
+            project = document.Project;
+        }
     }
 
     return project;
 }
```

Forcing use of updated project with normalized documents is possible by overriding `ValidateCodeFixTreeIsNormalized = false`, but it generally indicates a bug in the code fixer and/or underlying syntax generation/update APIs used by the fixers, which will prevent such a fixer from running properly in host or application that does not normalize the output of code fixers, as this is not a requirement placed upon any host by the Roslyn syntax generation API and tool chain.

Reply:
➡️ Moved to #140 rather than block this initial pull request